server/src/main/java/org/opensearch/indices/recovery/RecoveryListener.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.indices.recovery;
+
+import org.opensearch.OpenSearchException;
+import org.opensearch.cluster.routing.ShardRouting;
+import org.opensearch.indices.cluster.IndicesClusterStateService;
+import org.opensearch.indices.replication.common.ReplicationListener;
+import org.opensearch.indices.replication.common.ReplicationState;
+
+/**
+ * Listener that runs on changes in Recovery state
+ *
+ * @opensearch.internal
+ */
+public class RecoveryListener implements ReplicationListener {
+
+    /**
+     * ShardRouting with which the shard was created
+     */
+    private final ShardRouting shardRouting;
+
+    /**
+     * Primary term with which the shard was created
+     */
+    private final long primaryTerm;
+
+    private final IndicesClusterStateService indicesClusterStateService;
+
+    public RecoveryListener(
+        final ShardRouting shardRouting,
+        final long primaryTerm,
+        IndicesClusterStateService indicesClusterStateService
+    ) {
+        this.shardRouting = shardRouting;
+        this.primaryTerm = primaryTerm;
+        this.indicesClusterStateService = indicesClusterStateService;
+    }
+
+    @Override
+    public void onDone(ReplicationState state) {
+        RecoveryState RecState = (RecoveryState) state;
+        indicesClusterStateService.getShardStateAction()
```

Comment:
I think it is a better option to introduce `handleRecoveryDone` method into `indicesClusterStateService` and encapsulate this logic inside, it will also let us drop unnecessary `getShardStateActionListener()` and `getShardStateAction()` methods.